### PR TITLE
[ONNX] Extend ONNX Frontend with BlackmanWindow, HammingWindow and HannWindow operators

### LINKED ISSUE #18483: Extend ONNX Frontend with `BlackmanWindow`, `HammingWindow` and `HannWindow` operators
## Context
Neural networks are graphs consisting of nodes called operators. Each operator corresponds to a mathematical function, usually described in framework's documentation or an AI standard, such as ONNX.

OpenVINO ONNX Frontend is a component responsible for working with ONNX graphs and requires implementation of different ONNX operators in order to use ONNX models.

This task requires extending OpenVINO ONNX Frontend with `BlackmanWindow`, `HammingWindow` and `HannWindow` operators.

Necessary help will be provided by ONNX Fronted team.

## Operator specification
Operator details can be found in [ONNX Operators](https://github.com/onnx/onnx/blob/main/docs/Operators.md).

## Todo list
1. Create `.hpp` and `.cpp` files for *Windows [here](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op)
2. Prepare an implementation of this operator in form of a function. It should be placed in [opset 1 namespace](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/unique.hpp#L13).
3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560) while keeping alphabetical order
4. Create test model(s) in [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/). OpenVINO test infrastructure then converts prototxt files to ONNX models - you will use those models later in tests
5. Add tests covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import.in.cpp)

More details in [adding operators to ONNX Frontend guide](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md)

## Resources
- Guides on [ONNX Frontend architecture](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/architecture.md) and [ONNX Frontend tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/tests.md) are also available
- Contribution guidelines can be found [here](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- You can also see a [blog post on contributing to OpenVINO](https://medium.com/openvino-toolkit/how-to-contribute-to-an-ai-open-source-project-c741f48e009e)

## Example PRs
- https://github.com/openvinotoolkit/openvino/pull/15076
- https://github.com/openvinotoolkit/openvino/pull/14936
- https://github.com/openvinotoolkit/openvino/pull/12510

## Tips
In ONNX specification example code, you can find `numpy` implementation which can be a good reference. 
[BlackmanWindow](https://github.com/onnx/onnx/blob/main/docs/Operators.md#blackmanwindow)
```
# periodic window
size = np.int32(10)
a0 = 0.42
a1 = -0.5
a2 = 0.08
y = a0
y += a1 * np.cos(2 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / size)
y += a2 * np.cos(4 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / size)

# symmetric window
size = np.int32(10)
a0 = 0.42
a1 = -0.5
a2 = 0.08
y = a0
y += a1 * np.cos(
    2 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / (size - 1)
)
y += a2 * np.cos(
    4 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / (size - 1)
)
```
[HammingWindow](https://github.com/onnx/onnx/blob/main/docs/Operators.md#hammingwindow)
```
# periodic window
size = np.int32(10)
a0 = 25 / 46
a1 = 1 - a0
y = a0 - a1 * np.cos(
    2 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / size
)

#symmetric window
size = np.int32(10)
a0 = 25 / 46
a1 = 1 - a0
y = a0 - a1 * np.cos(
    2 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / (size - 1)
)
```
[HannWindow](https://github.com/onnx/onnx/blob/main/docs/Operators.md#hannwindow)
```
# periodic window
size = np.int32(10)
a0 = 0.5
a1 = 0.5
y = a0 - a1 * np.cos(
    2 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / size
)

# symmetric window
size = np.int32(10)
a0 = 0.5
a1 = 0.5
y = a0 - a1 * np.cos(
    2 * 3.1415 * np.arange(0, size, 1, dtype=np.float32) / (size - 1)
)
```
Note that OV equivalant of `numpy.arange` can be [OV Range op](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/generation/Range_4.md)

## Contact points
@mbencer
@p-wysocki
Don't hesitate to reach out, we're here to help!

### PR DESCRIPTION
**Pull request by siddhant-0707**
### Details:
 - created implementation for BlackmanWindow, HammingWindow and HannWindow operators
 - created basic tests for them

### Tickets:
 - Closes #18483

### CODE REVIEW

**Comment by siddhant-0707** [comment]
Hey @mbencer could you help me out with this? This is my first time contributing here.

**Comment by p-wysocki** [comment]
@siddhant-0707 there was an issue with CI on master, it has been fixed yesterday. Please update your branch, the checks failing on CMake step (all of them from what I see) will now start building and run actual tests. Sorry for the inconvenience. :)

**Comment by p-wysocki** [comment]
@siddhant-0707 do you have access to Azure pipelines logs? There's apparently some issue with building the new cpp files.

**Comment by siddhant-0707** [comment]
@p-wysocki Yep I can see the Azure pipelines logs. Will get back to you tomorrow.

**Review by gkrivor — CHANGES_REQUESTED** [review]
Great job! I think we should slightly polish the PR and merge.

**Comment by gkrivor** [comment]
@siddhant-0707 , btw, which code editor do you use? If it is MS VS - use "Ctrl + K, Ctrl + D" to fix coding style issues. Or thru menu Edit > Advanced > Format Document

**Comment by siddhant-0707** [comment]
Hey guys I made the requested changes but need some help with the coding style issues. I ran Format Document from the command palette (VSCode) and it seemed to work. Format looks same as other files in the repo but tests still failing.

File: src/frontends/onnx/frontend/src/op/hammingwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/hammingwindow.cpp:23]: You don't need this at all. get_ngraph_element_type returns element::Type or throw an exception if it isn't supported. So, you should use output_datatype directly instead of tensor_type.

File: src/frontends/onnx/frontend/src/op/hannwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/hannwindow.cpp:23]: Here is also you can use output_datatype directly

File: src/frontends/onnx/frontend/src/op/blackmanwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.cpp:101]: As I can see you don't need this constant, remove this line and use y_values as an output.

File: src/frontends/onnx/frontend/src/op/hammingwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/hammingwindow.cpp:92]: Please, check this place too, use y_values as an output

**Review by gkrivor — COMMENTED** [review]
I think you've successfully solved a code-style warnings.

**Comment by gkrivor** [comment]
Hi, @siddhant-0707! Great to see you are continue improving the PR! I'll try to build your change locally today or tomorrow, I think you should just use some additional macro to prevent deprecation warnings. But I need to clarify some points before.

File: src/frontends/onnx/frontend/src/op/blackmanwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.cpp:11]: Some parts of code are marked as deprecated due to upcoming refactoring. That's why I get an deprecation warnings (raised as errors in CI).

To prevent deprecated warnings please add OPENVINO_SUPPRESS_DEPRECATED_START before this line...

File: src/frontends/onnx/frontend/src/op/blackmanwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.cpp:67]: ... and OPENVINO_SUPPRESS_DEPRECATED_END after this line. You should do it for all *.hpp/*.cpp files.

As a reference you can take a look on https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/cast.cpp and https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/cast.hpp

I think it should be enough to build.

**Review by gkrivor — CHANGES_REQUESTED** [review]
Sorry for a delay, updating:

File: src/frontends/onnx/frontend/src/op/blackmanwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.cpp:33]: Looks like latest failures M_PI requires #include <math.h> (hope without additional definitions). Could you add, please?

**Review by gkrivor — CHANGES_REQUESTED** [review]
Let's take a look after adding the math.h header to files where you use M_PI, I may expect some further errors should disappear because I didn't see them locally. Or I'll check with CI team any difference with local build.

**Review by gkrivor — CHANGES_REQUESTED** [review]
Hi, finally I got a time to build your PR locally. Sorry, for a delay. I point on problems I fixed locally: added defines and "f" for float values. I built it on Windows, so if we face with issues on linux - need to return back again. I hope we are completing fix CI issues.

Btw, did you try to run tests which are available for added functionality? I mean existing tests like listed in:
https://github.com/openvinotoolkit/openvino/blob/7d1b7b6becf7b75d8d7f675d5262b1182235d90f/src/frontends/onnx/tests/tests_python/test_backend.py#L379
and
https://github.com/openvinotoolkit/openvino/blob/7d1b7b6becf7b75d8d7f675d5262b1182235d90f/src/bindings/python/tests_compatibility/test_onnx/test_backend.py#L470

**Comment by siddhant-0707** [comment]
i'll test them out today, thanks for pointing it out

File: src/frontends/onnx/frontend/src/op/blackmanwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.cpp:20]:
```suggestion
    const auto output_datatype = common::get_ngraph_element_type(node.get_attribute_value<int64_t>("output_datatype", 1));
```
I'm able to fix linkage issues on Linux by adding <int64_t> for get_attribute_value in each file, could you do the same, please?

File: src/frontends/onnx/frontend/src/op/blackmanwindow.hpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.hpp:9]: Could you align this part with other headers, like  https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/asinh.hpp

```
#pragma once

#include "openvino/core/deprecated.hpp"
OPENVINO_SUPPRESS_DEPRECATED_START
```
for all headers

**Review by gkrivor — CHANGES_REQUESTED** [review]
Hi! I became able to build your PR on Linux with listed changes, could you reflect them in your PR?

File: src/frontends/onnx/frontend/src/op/blackmanwindow.hpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.hpp:15]: Sorry, looks like I was misleading you. Because you are adding new operations, they should be in set_1, you did it right initially.

**Review by gkrivor — CHANGES_REQUESTED** [review]
OK, we've fixed build issues, let's fix test issues. Due to current implementation we need to use set_1 for new operations. I slightly missed this point.

File: src/frontends/onnx/frontend/src/op/blackmanwindow.hpp
Comment (siddhant-0707) [reply src/frontends/onnx/frontend/src/op/blackmanwindow.hpp]: No problem 😃

**Review by gkrivor — CHANGES_REQUESTED** [review]
OK, great, next issues looks like caused by your implementation. Test are in ov_onnx_frontend_tests binary. Could you fix them? I can point on wrong "size"-type (it is tensor, but in docs I see it is scalar).

File: src/frontends/onnx/tests/models/blackmanwindow_periodic.prototxt
Comment (siddhant-0707) [reply src/frontends/onnx/tests/models/blackmanwindow_periodic.prototxt:32]: Should I remove shape {} as well, or something else is required there?

Comment (gkrivor) [reply src/frontends/onnx/tests/models/blackmanwindow_periodic.prototxt]: I think you should leave it

**Comment by gkrivor** [comment]
Hi, great job! As I understand from logs - now you have issues around divide operation, it requires same arguments for calculations. You need to fix/align operand's types, or use Cast/CastLike to cast one type (int) to another type (float).

**Comment by gkrivor** [comment]
Hi, regarding issues, Cast is onnx operation, in OV it is Convert, could you try it?

**Comment by gkrivor** [comment]
I didn't dig deep into algorithm, why can't you use, for example, float (f32) in calculations? Current error also could be fixed by convert, but I'm not sure it is good choice...

**Comment by siddhant-0707** [comment]
Yes, you may be right. We can convert the final y_values to the output_datatype as the final step. Tests would still be required.

File: src/frontends/onnx/frontend/src/op/blackmanwindow.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/blackmanwindow.cpp:18]: Could you take a look on this modification of your code (I verified it locally) and do similar here and for other implementations:
```cpp
OutputVector blackmanwindow(const Node& node) {
    const auto size = node.get_ng_inputs().at(0);
    const auto output_datatype =
        common::get_ngraph_element_type(node.get_attribute_value<int64_t>("output_datatype", 1));
    const bool periodic = node.get_attribute_value<int64_t>("periodic", 1) == 1;

    const ov::PartialShape shape = size.get_partial_shape();
    const std::vector<size_t> axis_lengths = shape.to_shape();

    // Weights as described in ONNX BlackmanWindow docs
    // https://github.com/onnx/onnx/blob/main/docs/Operators.md#blackmanwindow
    const auto float_size = std::make_shared<default_opset::Convert>(size, ov::element::f32);
    const auto a_0 =
        std::make_shared<default_opset::Constant>(ov::element::f32, ov::Shape(), std::vector<float>{0.42f});
    const auto a_1 =
        std::make_shared<default_opset::Constant>(ov::element::f32, ov::Shape(), std::vector<float>{-0.50f});
    const auto a_2 =
        std::make_shared<default_opset::Constant>(ov::element::f32, ov::Shape(), std::vector<float>{0.08f});

    const auto start =
        std::make_shared<default_opset::Constant>(ov::element::f32, ov::Shape(), std::vector<float>{0.0f});
    const auto one_const =
        std::make_shared<default_opset::Constant>(ov::element::f32, ov::Shape(), std::vector<float>{1.0f});
    const auto two_const =
        std::make_shared<default_opset::Constant>(ov::element::f32, ov::Shape(), std::vector<float>{2.0f});
    const auto four_const =
        std::make_shared<default_opset::Constant>(ov::element::f32, ov::Shape(), std::vector<float>{4.0f});
    const auto range = std::make_shared<default_opset::Range>(start, size, one_const, ov::element::f32);
    const auto pi = default_opset::Constant::create(ov::element::f32, ov::Shape(), std::vector<float>{static_cast<float>(M_PI)});
    std::shared_ptr<ov::Node> factor_1, factor_2;
    if (periodic) {
        factor_1 = std::make_shared<default_opset::Multiply>(
            range,
            std::make_shared<default_opset::Divide>(std::make_shared<default_opset::Multiply>(pi, two_const),
                                                    float_size));
        factor_2 = std::make_shared<default_opset::Multiply>(
            range,
            std::make_shared<default_opset::Divide>(std::make_shared<default_opset::Multiply>(pi, four_const),
                                                    float_size));
    } else {
        factor_1 = std::make_shared<default_opset::Multiply>(
            range,
            std::make_shared<default_opset::Divide>(std::make_shared<default_opset::Multiply>(pi, two_const),
                                                    std::make_shared<default_opset::Subtract>(float_size, one_const)));
        factor_2 = std::make_shared<default_opset::Multiply>(
            range,
            std::make_shared<default_opset::Divide>(std::make_shared<default_opset::Multiply>(pi, four_const),
                                                    std::make_shared<default_opset::Subtract>(float_size, one_const)));
    }

    const auto cos_1 = std::make_shared<default_opset::Cos>(factor_1);
    const auto cos_2 = std::make_shared<default_opset::Cos>(factor_2);
    const auto scaled_cos_1 = std::make_shared<default_opset::Multiply>(cos_1, a_1);
    const auto scaled_cos_2 = std::make_shared<default_opset::Multiply>(cos_2, a_2);
    const auto y_values =
        std::make_shared<default_opset::Add>(std::make_shared<default_opset::Add>(a_0, scaled_cos_1), scaled_cos_2);

    if (output_datatype == element::f32) {
        return {y_values};
    } else {
        return {std::make_shared<default_opset::Convert>(y_values, output_datatype)};
    }
}
```

**Review by gkrivor — CHANGES_REQUESTED** [review]
Hi, I tried to verify your implementation, could you take a look on my comments and align your code? It should fix most issues.
I saw an issue on GPU, if you will see similar (or our CI will see it) - we may need to disable tests for the device (let's see it later).

File: src/frontends/onnx/tests/models/blackmanwindow_symmetric.prototxt
Comment (siddhant-0707) [reply src/frontends/onnx/tests/models/blackmanwindow_symmetric.prototxt:14]: Test for periodic case is [src/frontends/onnx/tests/models/blackmanwindow_periodic.prototxt](https://github.com/openvinotoolkit/openvino/blob/081cc1c278dc939da4511962cdb16a07b19496ad/src/frontends/onnx/tests/models/blackmanwindow_periodic.prototxt)

File: src/frontends/onnx/tests/onnx_import.in.cpp
Comment (siddhant-0707) [reply src/frontends/onnx/tests/onnx_import.in.cpp:6729]: Sorry but I'm unsure where to get the actual expected values.

File: src/frontends/onnx/frontend/src/op/blackmanwindow.cpp
Comment (siddhant-0707) [reply src/frontends/onnx/frontend/src/op/blackmanwindow.cpp:18]: Yep, looks good!

File: src/frontends/onnx/tests/models/blackmanwindow_symmetric.prototxt
Comment (gkrivor) [reply src/frontends/onnx/tests/models/blackmanwindow_symmetric.prototxt:14]: It works another way. To be "symmetric" it should be named "periodic" and has a 0 as value. For other cases your code is correct, mistake is only in this line.

File: src/frontends/onnx/tests/onnx_import.in.cpp
Comment (gkrivor) [inline src/frontends/onnx/tests/onnx_import.in.cpp:6775]: How did you get verification values? I generated slightly another set: {0.0869565f , 0.17414439f, 0.40240526f, 0.68455124f, 0.9128121f , 1.f, 0.9128121f, 0.6845511f, 0.4024052f, 0.17414442f}
I generate them by using python, for example:
```python
import numpy as np
np.set_printoptions(suppress=True,formatter=None,precision=16)
size = np.int32(10)
a0 = 25 / 46
a1 = 1 - a0
y = a0 - a1 * np.cos(2 * np.pi * np.arange(0, size, 1, dtype=np.float32) / size)
y
```
and it output:
array([0.0869565 , 0.17414439, 0.40240526, 0.68455124, 0.9128121 ,
       1.        , 0.9128121 , 0.6845511 , 0.4024052 , 0.17414442],
      dtype=float32)

**Review by gkrivor — CHANGES_REQUESTED** [review]
Looks like your solution is fully functional and you are able to remove corresponding tests from this file 
https://github.com/openvinotoolkit/openvino/blob/6bec4fc6dfae3929bfddc3bea8ff2a38c7f3f820/src/frontends/onnx/tests/tests_python/test_backend.py#L379

File: src/frontends/onnx/tests/onnx_import.in.cpp
Comment (gkrivor) [reply src/frontends/onnx/tests/onnx_import.in.cpp]: Arghhh, linux… try please std::string(“${BACKEND_NAME}”) != std::string(“IE_GPU”)
If not - I’ll take a look tomorrow

File: src/frontends/onnx/tests/onnx_import.in.cpp
Comment (gkrivor) [reply src/frontends/onnx/tests/onnx_import.in.cpp:6785]: Understand. Yes, you are right, just update data using np.pi. Looks like it is the last blocker. I hope.

**Comment by gkrivor** [comment]
Great job! Let's wait for CI results!

**Comment by gkrivor** [comment]
Congratulations with green CI! I'm waiting when our internal works will be finished and I'll run last pre-commit checks before merge.

**Review by gkrivor — APPROVED** [review]

**Comment by siddhant-0707** [comment]
Thank you so much for the help @gkrivor 😃

**Comment by p-wysocki** [comment]
That's an enormous piece of work. Thank you and congratulations @siddhant-0707!

**Pull request merged**
